src/reducer.ts

```diff
@@ -660,10 +660,17 @@ export default produce((draft: State, action: Action) => {
   }
 
   if (action.type === ActionType.LoadMessageArchive) {
-    action.messages.forEach((message) => {
-      draft.messages.entities[message.id] = message
-      draft.messages.ids.push(message.id)
-    })
+    const nextEntities = {
+      ...current(draft).messages.entities,
+      ...action.messages.reduce((entities, message) => {
+        entities[message.id] = message
+        return entities
+      }, {})
+    }
+
+    draft.messages.entities = nextEntities
+    draft.messages.ids = Object.keys(nextEntities)
```

Comment:
This converts from an array of message objects to the entity state style, while making sure that we don't inadvertently duplicate any ids … messages coming in through the action should overwrite any messages that already exist in state (if all their values are the same this will wind up being a no-op on the Immer side)